crates/exex/src/manager.rs

```diff
@@ -258,75 +260,114 @@ impl Future for ExExManager {
+
+                // it is a logic error for this to ever underflow since the manager manages the
+                // notification IDs
+                let notification_id = exex
+                    .next_notification_id
+                    .checked_sub(self.min_id)
+                    .expect("exex expected notification ID outside the manager's range");
+                if let Some(notification) = self.buffer.get(notification_id) {
+                    debug!(exex.id, notification_id, "sent notification to exex");
+                    if let Poll::Ready(Err(err)) = exex.send(notification) {
+                        // the channel was closed, which is irrecoverable for the manager
+                        return Poll::Ready(Err(err.into()))
+                    }
                 }
+                min_notification_id = min_notification_id.min(exex.next_notification_id);
+                self.exex_handles.push(exex);
             }
-        }
 
-        // update watch channel block number
-        let finished_height = self.exex_handles.iter_mut().try_fold(u64::MAX, |curr, exex| {
-            let height = match exex.finished_height {
-                None => return Err(()),
-                Some(height) => height,
-            };
-
-            if height < curr {
-                Ok(height)
-            } else {
-                Ok(curr)
+            // remove processed buffered notifications
+            self.buffer.retain(|&(id, _)| id >= min_notification_id);
+            self.min_id = min_notification_id;
+            debug!(min_notification_id, "lowest notification id in buffer updated");
+
+            // update capacity
+            self.update_capacity();
+
+            // handle incoming exex events
+            for exex in self.exex_handles.iter_mut() {
+                while let Poll::Ready(Some(event)) = exex.receiver.poll_recv(cx) {
+                    debug!(?event, id = exex.id, "received event from exex");
+                    exex.metrics.events_sent_total.increment(1);
+                    match event {
+                        ExExEvent::FinishedHeight(height) => exex.finished_height = Some(height),
+                    }
+                }
+            }
+
+            // update watch channel block number
+            let finished_height = self.exex_handles.iter_mut().try_fold(u64::MAX, |curr, exex| {
+                let height = match exex.finished_height {
+                    None => return Err(()),
+                    Some(height) => height,
+                };
+
+                if height < curr {
+                    Ok(height)
+                } else {
+                    Ok(curr)
+                }
+            });
+
+            if let Ok(finished_height) = finished_height {
+                let _ = self.finished_height.send(Some(finished_height));
+            }
+
+            // only return pending if the buffer is not empty
```

Comment:
this returns `Poll::Pending` if the buffer is empty, so the comment is not accurate

Reply:
Changed this to returning pending only when it is not empty and Poll:Ready(Ok()) otherwise